Tools/cases_generator/generate_cases.py

```diff
@@ -69,17 +67,45 @@ def always_exits(block: parser.Block) -> bool:
     return line.startswith(("goto ", "return ", "DISPATCH", "GO_TO_", "Py_UNREACHABLE()"))
 
 
-def write_instr(instr: InstDef, predictions: set[str], indent: str, f: TextIO, dedent: int = 0):
-    assert instr.block
+def find_cache_size(instr: InstDef, families: list[parser.Family]) -> str | None:
+    for family in families:
+        if instr.name == family.members[0]:
+            return family.size
+
+
+def write_instr(
+    instr: InstDef, predictions: set[str], indent: str, f: TextIO, dedent: int = 0, cache_size: str | None = None
+) -> int:
+    # Returns cache offset
     if dedent < 0:
         indent += " " * -dedent
+    # Separate stack inputs from cache inputs
+    input_names: set[str] = set()
+    stack: list[parser.StackEffect] = []
+    cache: list[parser.CacheEffect] = []
+    for input in instr.inputs:
+        if isinstance(input, parser.StackEffect):
+            stack.append(input)
+            input_names.add(input.name)
+        else:
+            assert isinstance(input, parser.CacheEffect), input
+            cache.append(input)
+    outputs = instr.outputs
+    cache_offset = 0
+    for ceffect in cache:
+        if ceffect.name != "unused":
+            bits = ceffect.size * 16
```

Comment:
Again, minor, but the connection between this magic constant and the width of a code unit isn't obvious when reading:

```suggestion
            bits = ceffect.size * CODEUNIT_BITS
```